File: mobile/android/app/src/main/java/com/fortis/businessportal/sdk/PaymentHandler.java
package com.sdkplayground.newarchitecture.sdk;


import android.Manifest;
import android.app.Activity;
import android.content.pm.PackageManager;

import com.facebook.react.bridge.ReactApplicationContext;
import com.idtechproducts.device.ReaderInfo;
import com.zeamster.paymentsdk.resthandler.ICallback;
import com.zeamster.paymentsdk.resthandler.RestServiceClient;
import com.zeamster.paymentsdk.service.TransactionService;
import com.zeamster.paymentsdk.transaction.EMVTransaction;
import com.zeamster.paymentsdk.transaction.ICallbackEMV;
import com.zeamster.paymentsdk.transaction.TransactionAction;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.Iterator;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

import androidx.core.content.ContextCompat;
import okhttp3.FormBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class PaymentHandler implements ICallback, ICallbackEMV {

    private final Activity activity;
    private final EventEmitter callbackUtils;

    private Result result;
    private RestServiceClient restServiceClient;
    private TransactionAction transactionAction;
    private EMVTransaction emvTransaction;
    private HashMap<String, Object> payload = new HashMap<>();
    private ReaderInfo.DEVICE_TYPE deviceType;
    private String transactionId = "";

    public PaymentHandler(final ReactApplicationContext reactContext, EventEmitter eventEmitter) {
        this.activity = reactContext.getCurrentActivity();
        this.callbackUtils = eventEmitter;
    }

    private boolean setup(JSONArray args) {

        if (args != null && args.length() > 0) {
            try {
                String arg = args.getString(0);
                if (arg != null && arg.length() > 0) {
                    JSONObject jsonObject = new JSONObject(arg);
                    JSONObject header = jsonObject.getJSONObject("header");
                    String validationStatus = validateAuthParameters(header);
                    if (validationStatus.equals("success")) {
                        restServiceClient = new RestServiceClient(header.optString("protocol"), header.optString("hostname"),
                                                                  header.optString("apiEndpoint"), this);
                        HashMap<String, String> requestHeader = new HashMap<>();
                        requestHeader.put("developer-id", header.optString("developerId"));
                        requestHeader.put("user-id", header.optString("userId"));

                        String authType = header.optString("authType");
                        requestHeader.put("authType", header.optString("authType"));
                        if (authType.equals("token")) {
                            if (header.has("accessToken")) {
                                requestHeader.put("access-token", header.optString("accessToken"));
                            } else {
                                // form parameters
                                RequestBody formBody = new FormBody.Builder()
                                        .add("username", header.optString("username"))
                                        .add("password", header.optString("password"))
                                        .add("domain", header.optString("domain"))
                                        .build();

                                Request request = new Request.Builder()
                                        .url(header.optString("protocol") + "://" + header.optString("hostname") + "/v2/token")
                                        .addHeader("developer-id", header.optString("developerId"))
                                        .addHeader("Content-Type", "application/json")
                                        .addHeader("Accept", "application/json")
                                        .post(formBody)
                                        .build();

                                OkHttpClient httpClient = new OkHttpClient();
                                try (Response response = httpClient.newCall(request).execute()) {

                                    if (!response.isSuccessful()) {
                                        throw new IOException("Unexpected code " + response);
                                    }

                                    // Get response body
                                    JSONObject tokenResponse = new JSONObject(response.body().string());
                                    JSONObject tokenBody = tokenResponse.getJSONObject("token");
                                    requestHeader.put("access-token", tokenBody.optString("token"));
                                } catch (Exception ex) {
                                    requestHeader.put("access-token", "dummy");
                                    callbackUtils.error(ex.getMessage());
                                }
                            }
                        } else if (authType.equals("hmac")) {
                            int currTime = Math.round((new Date()).getTime() / 1000);
                            String hashKey = header.optString("userHashKey");
                            String valueToDigest = header.optString("userId") + String.valueOf(currTime);
                            byte[] key = hashKey.getBytes();

                            String hashValue = generateHmac256(valueToDigest, key);

                            requestHeader.put("hash-key", hashValue);
                            requestHeader.put("timestamp", String.valueOf(currTime));
                        } else if (authType.equals("apikey")) {
                            requestHeader.put("user-api-key", header.optString("userAPIKey"));
                        }

                        restServiceClient.setRequestHeader(requestHeader);
                        if (header.has("transactionId")) {
                            transactionId = header.optString("transactionId");
                        }
                        if (header.has("deviceType")) {
                            deviceType = getDeviceType(header.optString("deviceType"));
                        }
                        if (jsonObject.has("queryParams")) {
                            restServiceClient.setQueryParams(toMap(jsonObject.getJSONObject("queryParams")));
                        }
                        if (jsonObject.has("body")) {
                            payload = toMap(jsonObject.getJSONObject("body"));
                        } else {
                            payload = new HashMap<>();
                        }
                        if (payload.size() > 0) {
                            if (payload.containsKey("action")) {
                                transactionAction = getTransactionAction(payload.get("action").toString());
                            }
                        }

                        return true;
                    } else {
                        callbackUtils.error(validationStatus);
                        return false;
                    }
                } else {
                    callbackUtils.error("Expected non-empty argument.");
                    return false;
                }
            } catch (Exception ex) {
                callbackUtils.error(ex.getMessage());
                return false;
            }
        } else {
            callbackUtils.error("Expected non-empty argument.");
            return false;
        }
    }

    private static String validateAuthParameters(JSONObject header) {
        if (header != null && header.has("protocol") && header.has("hostname") && header.has("apiEndpoint")) {
            if (header.has("authType")) {
                String authType = header.optString("authType");
                if (authType.equals("token")) {
                    if (header.has("developerId") && ((header.has("username") && header.has("password") && header.has("domain")) || header.has(
                            "accessToken"))) {
                        return "success";
                    } else {
                        return "Invalid Auth Parameters";
                    }
                } else if (authType.equals("hmac")) {
                    if (header.has("developerId") && header.has("userId") && header.has("userHashKey")) {
                        return "success";
                    } else {
                        return "Invalid Auth Parameters";
                    }
                } else if (authType.equals("apikey")) {
                    if (header.has("developerId") && header.has("userId") && header.has("userAPIKey")) {
                        return "success";
                    } else {
                        return "Invalid Auth Parameters";
                    }
                } else {
                    String returnVal = "Invalid Auth Type - " + authType;
                    return returnVal;
                }
            }
        }
        return "Invalid API Details";
    }

    private static String generateHmac256(
            String message,
            byte[] key) throws InvalidKeyException, NoSuchAlgorithmException {
        byte[] bytes = hmac("HmacSHA256", key, message.getBytes());
        return bytesToHex(bytes);
    }

    private static byte[] hmac(
            String algorithm,
            byte[] key,
            byte[] message) throws NoSuchAlgorithmException, InvalidKeyException {
        Mac mac = Mac.getInstance(algorithm);
        mac.init(new SecretKeySpec(key, algorithm));
        return mac.doFinal(message);
    }

    private static String bytesToHex(byte[] bytes) {
        final char[] hexArray = "0123456789abcdef".toCharArray();
        char[] hexChars = new char[bytes.length * 2];
        for (int j = 0, v; j < bytes.length; j++) {
            v = bytes[j] & 0xFF;
            hexChars[j * 2] = hexArray[v >>> 4];
            hexChars[j * 2 + 1] = hexArray[v & 0x0F];
        }
        return new String(hexChars);
    }

    private static HashMap<String, Object> toMap(JSONObject jsonObj) throws JSONException {
        HashMap<String, Object> map = new HashMap<>();
        Iterator<String> keys = jsonObj.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            Object value = jsonObj.get(key);
            if (value instanceof JSONArray) {
                value = toList((JSONArray) value);
            } else if (value instanceof JSONObject) {
                value = toMap((JSONObject) value);
            }
            map.put(key, value);
        }
        return map;
    }

    private static ArrayList<Object> toList(JSONArray array) throws JSONException {
        ArrayList<Object> list = new ArrayList<>();
        for (int i = 0; i < array.length(); i++) {
            Object value = array.get(i);
            if (value instanceof JSONArray) {
                value = toList((JSONArray) value);
            } else if (value instanceof JSONObject) {
                value = toMap((JSONObject) value);
            }
            list.add(value);
        }
        return list;
    }

    private TransactionAction getTransactionAction(String transactionAction) {
        switch (transactionAction.toLowerCase()) {
            case "sale":
                return TransactionAction.SALE;
            case "refund":
                return TransactionAction.REFUND;
            case "void":
                return TransactionAction.VOID;
            case "authonly":
                return TransactionAction.AUTHONLY;
            case "authcomplete":
                return TransactionAction.AUTHCOMPLETE;
            case "authincrement":
                return TransactionAction.AUTHINCREMENT;
            case "force":
                return TransactionAction.FORCE;
            case "tipadjust":
                return TransactionAction.TIPADJUST;
            case "debit":
                return TransactionAction.DEBIT;
            case "credit":
                return TransactionAction.CREDIT;
            case "edit":
                return TransactionAction.EDIT;
            case "avsonly":
                return TransactionAction.AVSONLY;
            case "view_transaction":
                return TransactionAction.VIEW_TRANSACTION;
            case "view_transactions_all":
                return TransactionAction.VIEW_TRANSACTIONS_ALL;
            case "view_transactions_filtered":
                return TransactionAction.VIEW_TRANSACTIONS_FILTERED;
            case "bininfo":
                return TransactionAction.BININFO;
            case "store":
                return TransactionAction.STORE;
        }
        return TransactionAction.UNKNOWN;
    }

    public void onSuccess(String response) {
        callbackUtils.success(response);
    }

    public void onFailure(String response) {
        callbackUtils.error(response);
    }

    private void updatePayload(JSONArray args) {

        if (args != null && args.length() > 0) {
            try {
                String arg = args.getString(0);
                if (arg != null && arg.length() > 0) {
                    JSONObject jsonObject = new JSONObject(arg);
                    if (jsonObject.has("body")) {
                        payload = toMap(jsonObject.getJSONObject("body"));
                    }
                } else {
                    callbackUtils.error("Expected non-empty argument.");
                }
            } catch (Exception ex) {
                callbackUtils.error(ex.getMessage());
            }
        } else {
            callbackUtils.error("Expected non-empty argument.");
        }
    }

    public void performTransaction(JSONArray args) {
        if (!setup(args)) {
            return;
        }

        TransactionService transactionService = new TransactionService(restServiceClient);
        transactionService.processTransaction(transactionAction, payload, transactionId);
    }

    public void performEMVTransaction(JSONArray args) {
        setup(args);
        updatePayload(args);
        emvTransaction.performEMVSale(transactionAction, payload);
    }

    private ReaderInfo.DEVICE_TYPE getDeviceType(String deviceType) {
        switch (deviceType) {
            case "IDTECH-VP3300BT":
                return ReaderInfo.DEVICE_TYPE.DEVICE_VP3300_BT;
            case "IDTECH-VP8800BT":
                return ReaderInfo.DEVICE_TYPE.DEVICE_VP8800_BT;
        }
        return ReaderInfo.DEVICE_TYPE.DEVICE_UNKNOWN;
    }

    public void setDeviceType(JSONArray args) throws JSONException {
        String deviceType = args.getString(0);
        emvTransaction.setDeviceType(getDeviceType(deviceType));
        callbackUtils.success("Success.");
    }

    public void scanForDevices(JSONArray args) {
        result = new Result(Result.Status.OK, "Scanning in progress..");
        callbackUtils.result(result);
        setup(args);
        emvTransaction = new EMVTransaction(restServiceClient, this, activity, deviceType);
        emvTransaction.isPluginRequest(true);

        // Previous version of the Cordova plugin used to request the permission
        if (ContextCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED
                || ContextCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            callbackUtils.error("Location permission not granted");
            return;
        }

        try {
            if (args != null && args.length() == 3 && !args.get(1).toString().equals("null") && !args.get(2).toString().equals("null")) {

                String searchString = args.getString(1);
                long timeout = Long.parseLong(args.getString(2));
                emvTransaction.scanForDevices(searchString, timeout);
            } else if (args != null && args.length() == 3 && !args.get(1).toString().equals("null")) {

                long timeout = Long.parseLong(args.getString(1));
                emvTransaction.scanForDevices(timeout);

            } else {
                emvTransaction.scanForDevices();
            }
        } catch (Exception ex) {
            callbackUtils.error(ex.getMessage());
        }
    }

    public void initialDeviceSetup(JSONArray args) {
        result = new Result(Result.Status.OK, "Running initial device setup..");
        callbackUtils.result(result);
        setup(args);
        emvTransaction.restServiceClient = restServiceClient;
        emvTransaction.runInitialDeviceSetup();
    }

    public void stopScanForDevices() {
        emvTransaction.stopScanForDevices();
        callbackUtils.success("Scanning Stopped.");
    }

    public void connectDeviceByName(JSONArray args) throws JSONException {
        String deviceName = args.getString(0);
        emvTransaction.connectDeviceByName(deviceName);
    }

    private void connectDeviceByAddress(JSONArray args) throws JSONException {
        String deviceAddress = args.getString(0);
        emvTransaction.connectDeviceByAddress(deviceAddress);
    }

    public void deviceMessage(String message) {
        result = new Result(Result.Status.OK, "deviceMessage!@#$" + message);
        callbackUtils.result(result);
    }

    public void deviceScanResponse(
            String deviceId,
            String deviceName) {

        result = new Result(Result.Status.OK, "deviceScanResponse!@#$"
                + (deviceId != null ? deviceId : "") + "!@#$" + (deviceName != null ? deviceName : ""));
        callbackUtils.result(result);
    }

    public void deviceScanCompleted() {
        result = new Result(Result.Status.OK, "deviceScanCompleted!@#$Scanning Completed");
        callbackUtils.result(result);
    }

    public void deviceConnected(String status) {
        result = new Result(Result.Status.OK, "deviceConnected!@#$" + status);
        callbackUtils.result(result);
    }

    public void deviceDisconnected(String status) {
        result = new Result(Result.Status.OK, "deviceDisconnected!@#$" + status);
        callbackUtils.result(result);
    }

    public void outputLogs(String log) {
        result = new Result(Result.Status.OK, "outputLogs!@#$" + log);
        callbackUtils.result(result);
    }

    public void disconnectDevice() {
        emvTransaction.disconnectDevice();
    }

    public void setTimeouts(JSONArray args) {
        try {
            if (args != null && args.length() == 2 && !args.get(0).toString().equals("null") && !args.get(1).toString().equals("null")) {
                int idleTimeout = Integer.parseInt(args.getString(0));
                int sleepTimeout = Integer.parseInt(args.getString(1));
                emvTransaction.setTimeouts(idleTimeout, sleepTimeout);
            }
        } catch (JSONException e) {
            callbackUtils.error(e.toString());
        }
    }
}

File: mobile/ios/Handlers/PaymentHandler.swift
import CryptoKit
import Foundation
import PaymentSDK

final class PaymentHandler: NSObject {
    
    private let callback: EventEmitter
    private let decoder: ArgumentsDecoder
    
    private(set) var emvTransaction: EMVTransaction?
    private(set) var restServiceClient: RestServiceClient?
    private(set) var transactionAction: TransactionAction?
    private(set) var payload: [String: String]?
    private(set) var transactionId: String = ""
    private(set) var deviceType: DeviceType = .UNKNOWN
    private(set) var connectingDevice: String = ""
    
    init(callback: EventEmitter, decoder: ArgumentsDecoder = .init()) {
        self.callback = callback
        self.decoder = decoder
    }
    
    deinit {
        emvTransaction = nil
        restServiceClient = nil
        transactionAction = nil
        payload = nil
        transactionId = ""
        deviceType = .UNKNOWN
        connectingDevice = ""
    }
    
    func performTransaction(args: String) throws {
        do {
            try setup(with: args)
            
            guard let restClient = restServiceClient, let transaction = transactionAction, let payload = payload else {
                throw(PaymentHandlerError.invalidParameters(action: "performTransaction"))
            }
            
            let paramMap = NSMutableDictionary(dictionary: payload)
            
            let transactionService = TransactionService(restClient: restClient)
            transactionService.processTransaction(transactionAction: transaction, paramMap: paramMap) { [self] responseData in
                if responseData.isEmpty {
                    callback.sendEvent(type: .error)
                } else {
                    callback.sendEvent(type: .ok, message: responseData)
                }
            }
        }
    }
    
    func performEMVTransaction(args: String) throws {
        do {
            try setup(with: args)
            
            guard let payload = payload else {
                throw(PaymentHandlerError.invalidParameters(action: "performEMVTransaction"))
            }
            
            if let result = emvTransaction?.PerformEMVSale(jsonPayload: payload), result {
                callback.sendEvent(type: .ok, message: "EMV Transaction Started...")
            } else {
                callback.sendEvent(type: .error)
            }
        }
    }
    
    func setDeviceType(args: String) throws {
        do {
            let device: [String] = try decoder.decode(args: args)
            
            guard let result = emvTransaction?.setDeviceType(type: getDeviceType(with: device.first)), result else {
                throw(PaymentHandlerError.invalidParameters(action: "setDeviceType"))
            }
            callback.sendEvent(type: .ok)
        }
    }
    
    func scanForDevices(args: String) throws {
        callback.sendEvent(type: .ok, message: "Scanning in progress...")
        
        do {
            try setup(with: args)
            
            _ = emvTransaction?.scanForDevices()
        }
    }
    
    func stopScanForDevices() throws {
        callback.sendEvent(type: .ok, message: "Stop scanning is not supported")
    }
    
    func connectDeviceByName(args: String) throws {
        do {
            let arguments: [String] = try decoder.decode(args: args)
            
            guard let device = arguments.first, let result = emvTransaction?.connectDeviceByName(name: device), result else {
                throw(PaymentHandlerError.invalidParameters(action: "connectDeviceByName"))
            }
            
            callback.sendEvent(type: .ok)
        }
    }
    
    func initialDeviceSetup(args: String) throws {
        callback.sendEvent(type: .ok, message: "Running initial device setup...")
        
        do {
            try setup(with: args)
            
            guard let restClient = restServiceClient else {
                throw(PaymentHandlerError.invalidParameters(action: "initialDeviceSetup"))
            }
            
            emvTransaction?.restServiceClient = restClient
            _ = emvTransaction?.initialDeviceSetup()
        }
    }
    
    func disconnectDevice() throws {
        _ = emvTransaction?.DisconnectDevice()
    }
    
    func setTimeouts(args: String) throws {
        do {
            let arguments: [String] = try decoder.decode(args: args)
            
            guard !arguments.isEmpty, arguments.count == 2, arguments[0] != "null", arguments[1] != "null" else {
                throw(PaymentHandlerError.invalidParameters(action: "setTimeouts"))
            }
            
            let idleTimeout = Int(arguments[0]) ?? 0
            let sleepTimeout = Int(arguments[1]) ?? 0
            _ = emvTransaction?.setTimeouts(idleMinutes: idleTimeout, sleepMinutes: sleepTimeout)
        }
    }
}

// MARK: - Private Methods

extension PaymentHandler {
    
    private func setup(with args: String) throws {
        let parameters: Transaction = try decoder.decode(args: args)
        
        guard let header = parameters.header else {
            throw(PaymentHandlerError.emptyArguments)
        }
        
        guard let httpProtocol = header.httpProtocol, let hostname = header.hostname, let apiEndpoint = header.apiEndpoint else {
            throw(PaymentHandlerError.invalidApiDetails)
        }
        
        guard let authType = header.authType else {
            throw(PaymentHandlerError.invalidAuthType)
        }
        
        guard let developerId = header.developerId, let userId = header.userId else {
            throw(PaymentHandlerError.invalidAuthParameters)
        }
        
        restServiceClient = RestServiceClient(apiProtocol: httpProtocol, apiHostName: hostname, apiEndpointPath: apiEndpoint)
        
        var requestHeader: [String: String] = [:]
        requestHeader["developer-id"]  = developerId
        requestHeader["user-id"] = userId
        requestHeader["authType"] = authType.rawValue
        
        switch authType {
        case .token:
            try setRequestHeaderByToken(with: header, requestHeader: &requestHeader)
        case .hmac:
            try setRequestHeaderByHMAC(with: header, requestHeader: &requestHeader)
        case .apikey:
            try setRequestHeaderByKey(with: header, requestHeader: &requestHeader)
        }
        
        _ = restServiceClient?.setHTTPRequestHeaders(headerParams: NSDictionary(dictionary: requestHeader))
        transactionId = header.transactionId ?? ""
        
        if let device = header.deviceType {
            deviceType = getDeviceType(with: device)
        } else {
            deviceType = getDeviceType(with: "IDTECH-VP3300BT")
        }
        
        if let body = parameters.body, let action = body.action {
            transactionAction = getTransactionAction(with: action)
            payload = try body.asDictionary()
        }
        
        if let restServiceClient {
            emvTransaction = EMVTransaction(restClient: restServiceClient)
            emvTransaction?.isPluginRequest = true
            emvTransaction?.pluginDelegate = self
            _ = emvTransaction?.setDeviceType(type: deviceType)
        }
    }
    
    private func setRequestHeaderByToken(with header: Transaction.Header, requestHeader: inout [String: String]) throws {
        if let accessToken = header.accessToken {
            requestHeader["access-token"] = accessToken
            return
        }
        
        guard let developerId = header.developerId, let httpProtocol = header.httpProtocol, let hostname = header.hostname else {
            throw(PaymentHandlerError.invalidAuthParameters)
        }
        
        var errorToReturn: Error?
        var accessToken: String?
        
        let semaphore = DispatchSemaphore(value: 1)
        
        if let url = URL(string: "\(httpProtocol)://\(hostname)/v2/token") {
            var request = URLRequest(url: url)
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.setValue("application/json", forHTTPHeaderField: "Accept")
            request.setValue(developerId, forHTTPHeaderField: "developer-id")
            request.httpMethod = "POST"
            
            var bodyParameters: [String: String?] = [:]
            bodyParameters["username"] = header.username
            bodyParameters["password"] = header.password
            bodyParameters["domain"] = header.domain
            
            let jsonData = try? JSONSerialization.data(withJSONObject: bodyParameters)
            request.httpBody = jsonData
            
            let session = URLSession.shared
            let task = session.dataTask(with: request) { data, response, error in
                defer { semaphore.signal() }
                
                var responseDictionary: [AnyHashable: Any]? = nil
                
                do {
                    if let data {
                        responseDictionary = try JSONSerialization.jsonObject(with: data, options: []) as? [AnyHashable: Any]
                    }
                } catch let parseError {
                    errorToReturn = parseError
                }
                
                let tokenObj = responseDictionary?["token"] as? [AnyHashable: Any]
                accessToken = tokenObj?["token"] as? String
            }
            task.resume()
            semaphore.wait()
            
            guard let accessToken else {
                throw(errorToReturn ?? PaymentHandlerError.invalidAuthParameters)
            }
            
            requestHeader["access-token"] = accessToken
        }
    }
    
    private func setRequestHeaderByHMAC(with header: Transaction.Header, requestHeader: inout [String: String]) throws {
        guard let userId = header.userId, let userHashKey = header.userHashKey else {
            throw(PaymentHandlerError.invalidAuthParameters)
        }
        
        let timeInSeconds = Int(Date().timeIntervalSince1970)
        let timestamp = String(format: "%1d", timeInSeconds)
        
        let hmacData = "\(userId)\(timestamp)"
        let key = SymmetricKey(data: Data(hmacData.utf8))
        
        let signature = HMAC<SHA256>.authenticationCode(for: Data(userHashKey.utf8), using: key)
        let userHash = Data(signature).map { String(format: "%02x", $0) }.joined()
        
        requestHeader["hash-key"] =  userHash
        requestHeader["timestamp"] = timestamp
    }
    
    private func setRequestHeaderByKey(with header: Transaction.Header, requestHeader: inout [String: String]) throws {
        guard let userAPIKey = header.userAPIKey else {
            throw(PaymentHandlerError.invalidAuthParameters)
        }
        
        requestHeader["user-api-key"] = userAPIKey
    }

    private func getDeviceType(with deviceType: String?) -> DeviceType {
        switch deviceType {
        case "IDTECH-VP3300BT":
            return .IDTECH_VP3300BT
        case "IDTECH-VP8800":
            return .IDTECH_VP8800
        default:
            return .UNKNOWN
        }
    }
    
    private func getTransactionAction(with action: String) -> TransactionAction {
        switch action.lowercased() {
        case "sale": return .Sale
        case "refund": return .Refund
        case "void": return .Void
        case "authonly": return .AuthOnly;
        case "authcomplete": return .AuthComplete
        case "authincrement": return .AuthIncrement
        case "force": return .Force
        case "tipadjust": return .TipAdjust
        case "debit": return .Debit
        case "credit": return .Credit
        case "edit": return .Edit
        case "avsonly": return .AVSOnly
        case "view_transaction": return .ViewRecord
        case "view_transactions_all": return .ViewRecordList
        case "bininfo": return .GetBinInfo
        case "store": return .Store
        default: return .NotSelected
        }
    }
}

// MARK: - PluginDelegate

extension PaymentHandler: PluginDelegate {
    
    func deviceMessage(_ message: String!) {
        callback.sendEvent(
            type: .ok,
            message: "\("deviceMessage!@#$")\(message ?? "")"
        )
    }
    
    func deviceScanResponse(_ deviceID: String!, deviceName: String!) {
        callback.sendEvent(
            type: .ok,
            message: "\("deviceScanResponse!@#$")\(deviceID ?? "")\("!@#$")\(deviceName ?? "")"
        )
    }
    
    func deviceConnected() {
        callback.sendEvent(
            type: .ok,
            message: "\("deviceConnected!@#$")\(connectingDevice)\(" - Connected")"
        )
    }
    
    func deviceDisconnected() {
        callback.sendEvent(
            type: .ok,
            message: "\("deviceDisconnected!@#$")\(connectingDevice)\(" - Disconnected")"
        )
    }
    
    func outputLogs(_ logs: String!) {
        callback.sendEvent(
            type: .ok,
            message: "\("outputLogs!@#$")\(logs ?? "")"
        )
    }
    
    func transactionResponse(_ data: String!) {
        callback.sendEvent(
            type: .ok,
            message: "\("transactionResponse!@#$")\(data ?? "")"
        )
    }
}
